docs/decisions/0014-custom-prompt-template-formats.md

```diff
@@ -0,0 +1,290 @@
+                    writer.WriteSafeString(result.GetValue<string>());
+                });
+        }
+
+        var template = handlebars.Compile(templateText);
+
+        var prompt = template(context.Variables);
+
+        return await Task.FromResult(prompt).ConfigureAwait(true);
+    }
+}
+```
+
+**Note: This is just a prototype implementation for illustration purposes only.**
+
+Some issues:
+
+1. The `IPromptTemplate` interface is not used and causes confusion.
+1. There is no way to allow developers to support multiple prompt template formats at the same time.
+
+
+There is one implementation of `IPromptTemplate` provided in the Semantic Kernel core package.
+The `RenderAsync` implementation just delegates to the `IPromptTemplateEngine`.
+The `Parameters` list get's populated with the parameters defined in the `PromptTemplateConfig` and any missing variables defined in the template.
+
+#### Handlebars Considerations
+
+Handlebars does not support dynamic binding of helpers. Consider the following snippet:
+
+```csharp
+HandlebarsHelper link_to = (writer, context, parameters) =>
+{
+    writer.WriteSafeString($"<a href='{context["url"]}'>{context["text"]}</a>");
+};
+
+string source = @"Click here: {{link_to}}";
+
+var data = new
+{
+    url = "https://github.com/rexm/handlebars.net",
+    text = "Handlebars.Net"
+};
+
+// Act
+var handlebars = HandlebarsDotNet.Handlebars.Create();
+handlebars.RegisterHelper("link_to", link_to);
+var template = handlebars1.Compile(source);
+// handlebars.RegisterHelper("link_to", link_to); This also works
+var result = template1(data);
+```
+
+Handlebars allows the helpers to be registered with the `Handlebars` instance either before or after a template is compiled.
+The optimum would be to have a shared `Handlebars` instance for a specific collection of functions and register the helpers just once.
+For use cases where the Kernel function collection may have been mutated we will be forced to create a `Handlebars` instance at render time
+and then register the helpers. This means we cannot take advantage of the performance improvement provided by compiling the template.
+
+## Decision Drivers
+
+In no particular order:
+
+* Support creating a semantic function without a `IKernel`instance.
+* Support late binding of functions i.e., having functions resolved when the prompt is rendered.
+* Support allowing the prompt template to be parsed (compiled) just once to optimize performance if needed.
+* Support using multiple prompt template formats with a single `Kernel` instance.
+* Provide simple abstractions which allow third parties to implement support for custom prompt template formats.
+
+## Considered Options
+
+* Obsolete `IPromptTemplateEngine` and replace with `IPromptTemplateFactory`.
+* 
+
+### Obsolete `IPromptTemplateEngine` and replace with `IPromptTemplateFactory`
+
+<img src="./diagrams/prompt-template-factory.png" alt="ISKFunction class relationships"/>
```

Comment:
Mark, you can add the Mermaid directly to the md file without needing to create and mmd and a png for it. 😊

```suggestion
```mermaid
classDiagram
    %% Use https://mermaid.live/ to preview this diagram. The VS Code extension does not handle namespaces.
    direction BT
    namespace SemanticKernel {
        class IPromptTemplate {
            <<interface>>
            IReadOnlyList<ParameterView> Parameters;
            public Task<string> RenderAsync(SKContext executionContext, CancellationToken cancellationToken = default);
        }

        class IPromptTemplateFactory {
            <<interface>>
            IPromptTemplate CreatePromptTemplate(string template, PromptTemplateConfig config)
        }
    }

    namespace TemplateEngine_Basic {
        class BasicPromptTemplate
        class FStringPromptTemplate
        class BasicPromptTemplateFactory
    }

    namespace TemplateEngine_Handlebars {
        class HandlebarsPromptTemplate
        class HandlebarsPromptTemplateFactory
    }

    IPromptTemplateFactory <|.. BasicPromptTemplateFactory
    IPromptTemplateFactory <|.. HandlebarsPromptTemplateFactory

    IPromptTemplate <|.. BasicPromptTemplate
    IPromptTemplate <|.. FStringPromptTemplate
    IPromptTemplate <|.. HandlebarsPromptTemplate
```
```

Reply:
This diagram won't render in VS Code, I had to use the online editor to create the diagram.

Also I tried using this in preview on GH web and it didn't work there either.